packages/swing-store/src/bundleStore.js

```diff
@@ -0,0 +1,242 @@
+  db.exec(`
+    CREATE TABLE IF NOT EXISTS bundles (
+      bundleID TEXT,
+      bundle BLOB,
+      PRIMARY KEY (bundleID)
+    )
+  `);
+
+  function bundleArtifactName(bundleID) {
+    return `bundle.${bundleID}`;
+  }
+
+  function bundleIdFromHash(hash) {
+    return `b1-${hash}`;
+  }
+
+  const sqlAddBundle = db.prepare(`
+    INSERT OR REPLACE INTO bundles
+      (bundleID, bundle)
+    VALUES (?, ?)
+  `);
+
+  /**
+   * Store a bundle.  Here the bundle itself is presumed valid.
+   *
+   * @param {string} bundleID
+   * @param {Bundle} bundle
+   */
+  function addBundle(bundleID, bundle) {
+    const { moduleFormat, endoZipBase64, endoZipBase64Sha512 } = bundle;
+    moduleFormat === 'endoZipBase64' ||
+      Fail`unsupported module format ${q(moduleFormat)}`;
+    bundleID === bundleIdFromHash(endoZipBase64Sha512) ||
+      Fail`bundleID ${q(bundleID)} does not match bundle`;
+    ensureTxn();
+    sqlAddBundle.run(bundleID, decodeBase64(endoZipBase64));
+    noteExport(bundleArtifactName(bundleID), bundleID);
+  }
+
+  const sqlHasBundle = db.prepare(`
+    SELECT count(*)
+    FROM bundles
+    WHERE bundleID = ?
+  `);
+  sqlHasBundle.pluck(true);
+
+  function hasBundle(bundleID) {
+    const count = sqlHasBundle.get(bundleID);
+    return count !== 0;
+  }
+
+  const sqlGetBundle = db.prepare(`
+    SELECT bundle
+    FROM bundles
+    WHERE bundleID = ?
+  `);
+  sqlGetBundle.pluck(true);
+
+  function getBundle(bundleID) {
+    bundleID.startsWith('b1-') || Fail`invalid bundleID ${q(bundleID)}`;
+    const rawBundle = sqlGetBundle.get(bundleID);
+    rawBundle || Fail`bundle ${q(bundleID)} not found`;
+    return harden({
+      moduleFormat: 'endoZipBase64',
+      endoZipBase64Sha512: bundleID.substring(3),
+      endoZipBase64: encodeBase64(rawBundle),
+    });
+  }
+
+  const sqlDeleteBundle = db.prepare(`
+    DELETE FROM bundles
+    WHERE bundleID = ?
+  `);
+
+  function deleteBundle(bundleID) {
+    if (hasBundle(bundleID)) {
+      ensureTxn();
+      sqlDeleteBundle.run(bundleID);
+      noteExport(bundleArtifactName(bundleID), undefined);
+    }
+  }
+
+  /**
+   * Read a bundle and return it as a stream of data suitable for export to
+   * another store.
+   *
+   * Bundle artifact names should be strings of the form:
+   *   `bundle.${bundleID}`
+   *
+   * @param {string} name
+   *
+   * @yields {Uint8Array}
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  async function* exportBundle(name) {
+    typeof name === 'string' || Fail`artifact name must be a string`;
+    const parts = name.split('.');
+    const [type, bundleID] = parts;
+    // prettier-ignore
+    (parts.length === 2 && type === 'bundle') ||
+      Fail`expected artifact name of the form 'bundle.{bundleID}', saw ${q(name)}`;
+    const bundle = getBundle(bundleID);
+    bundle || Fail`bundle ${q(name)} not available`;
+    yield* Readable.from(Buffer.from(bundle.endoZipBase64));
```

Comment:
I think the export artifacts should be binary blobs, not base64-encoded. The middleman accepting data from the exporter is expected to handle bytes, might as well take advantage of it and make the artifacts that much smaller.